sdk/storage/Azure.Storage.Blobs/src/BlobClient.cs

```diff
@@ -219,6 +219,15 @@ protected virtual BlobClient WithSnapshotImpl(string snapshot)
                     // allow retrying if it fails.
                     response.Value.Content = RetriableStream.Create(
                         response.GetRawResponse(),
+                         startOffset =>
+                            (this.StartDownloadAsync(
+                                    range,
+                                    accessConditions,
+                                    rangeGetContentHash,
+                                    startOffset,
+                                    cancellation)
+                                .ConfigureAwait(false).GetAwaiter().GetResult())
+                            .GetRawResponse(),
```

Comment:
RetriableStream is a pretty advanced scenario.  Would it be better to just have a single response factory that took a `bool async` instead of duplicating all of this?

If folks think that's too difficult for most users to get right, I'm fine having the duplication and can just write my own shared implementation that both response factories use if I really want it.

Reply:
Why do you think it would be better with the `bool`? Method calls are currently duplicated because there is no sync version of `StartDownloadAsync`. If there was one the code would've looked like this:

``` C#
 RetriableStream.Create(
                        response.GetRawResponse(),
                         startOffset =>
                            (this.StartDownload(
                                    range,
                                    accessConditions,
                                    rangeGetContentHash,
                                    startOffset,
                                    cancellation))
                            .GetRawResponse(),
                        async startOffset =>
                            (await this.StartDownloadAsync(
                                range,
                                accessConditions,
                                rangeGetContentHash,
                                startOffset,
                                cancellation)
                                .ConfigureAwait(false))
                            .GetRawResponse(),
                        // TODO: For now we're using the default ResponseClassifier
                        // on BlobConnectionOptions so we'll do the same here
                        new ResponseClassifier(),
                        Constants.MaxReliabilityRetries);
```

Reply:
> Why do you think it would be better with the `bool`?

No copies are better than some copies.

I think `StartDownloadAsync` should also take a `bool async` parameter as well.  😃  I'm pretty sure we'll end up doing that across the board to have single implementations of all the operations.  It's fine if we want to leave that as an implementation choice for client library authors and not force it in `RetriableStream`'s public API.

Reply:
Let's see how sync support pans out in large client libraries, you may be right that it would end up being `bool sync` everywhere but I'd like to avoid forcing it from Azure.Core side